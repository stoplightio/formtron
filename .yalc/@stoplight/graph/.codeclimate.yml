version: "2"
checks:
  argument-count:
    config:
      threshold: 6
  complex-logic:
    config:
      threshold: 10
  file-lines:
    config:
      threshold: 1000
  method-complexity:
    config:
      threshold: 30
  method-count:
    config:
      threshold: 30
  method-lines:
    config:
      threshold: 200
  nested-control-flow:
    config:
      threshold: 8
  return-statements:
    config:
      threshold: 5
  similar-code:
    config:
      threshold: 200
exclude_patterns:
  - "**/__tests__/"
