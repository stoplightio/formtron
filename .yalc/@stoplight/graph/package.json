{
  "name": "@stoplight/graph",
  "version": "0.0.0-48b7f808",
  "description": "Nodes'n Things.",
  "keywords": [],
  "main": "src/index.ts",
  "sideEffects": false,
  "homepage": "https://github.com/stoplightio/graph",
  "bugs": "https://github.com/stoplightio/graph/issues",
  "author": "Stoplight <support@stoplight.io>",
  "repository": {
    "type": "git",
    "url": "https://github.com/stoplightio/graph.git"
  },
  "license": "UNLICENSED",
  "files": [
    "**/*"
  ],
  "engines": {
    "node": ">=8.3.0"
  },
  "scripts": {
    "build": "sl-scripts build",
    "build.docs": "sl-scripts build:typedoc",
    "commit": "git-cz",
    "lint": "sl-scripts lint",
    "lint.fix": "yarn lint --fix",
    "release": "sl-scripts release",
    "release.docs": "sl-scripts release:docs",
    "release.dryRun": "sl-scripts release --dry-run --debug",
    "test": "sl-scripts test",
    "test.prod": "yarn lint && yarn test --coverage --no-cache --maxWorkers 2",
    "test.update": "yarn test --updateSnapshot",
    "test.watch": "yarn test --watch"
  },
  "dependencies": {
    "@stoplight/json": "1.1.x",
    "@stoplight/types": "2.x.x",
    "@stoplight/yaml": "1.0.x",
    "graphlib": "2.1.5",
    "lodash": "4.x.x",
    "urijs": "1.19.x",
    "vfile": "3.0.x"
  },
  "devDependencies": {
    "@stoplight/scripts": "3.0.2",
    "@stoplight/test-utils": "0.0.1",
    "@types/graphlib": "2.1.4",
    "@types/lodash": "4.x.x",
    "@types/swagger-schema-official": "2.0.x",
    "@types/urijs": "1.15.x",
    "@types/vfile": "2.2.x",
    "openapi3-ts": "1.1.x",
    "typescript": "3.1.6"
  },
  "lint-staged": {
    "*.{ts,tsx}": [
      "yarn lint.fix",
      "git add"
    ]
  },
  "husky": {
    "hooks": {
      "commit-msg": "commitlint -E HUSKY_GIT_PARAMS",
      "pre-commit": "lint-staged"
    }
  },
  "config": {
    "commitizen": {
      "path": "node_modules/cz-conventional-changelog"
    }
  },
  "commitlint": {
    "extends": [
      "@commitlint/config-conventional"
    ]
  },
  "release": {
    "extends": "@stoplight/scripts/release"
  }
}
