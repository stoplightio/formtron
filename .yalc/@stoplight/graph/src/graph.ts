import * as graphlib from 'graphlib';

import { IGraphHook, IGraphLoader, IGraphPlugin, INodeInstance, INodeOptions } from './types';

export class Graph {
  public readonly graph = new graphlib.Graph({
    directed: false,
    compound: true,
    multigraph: false,
  });

  /**
   * Nodes
   */

  get nodes(): INodeInstance[] {
    return this.graph.nodes().map(nodeId => this.graph.node(nodeId) as INodeInstance);
  }

  get edges() {
    return this.graph.edges();
  }

  public getNode<TNode extends INodeInstance>(id: string): TNode | undefined {
    return this.graph.node(id);
  }

  public createNode = async <Input extends INodeOptions, Output extends INodeInstance>(
    nodeProps: Input,
    parent?: INodeInstance
  ): Promise<Output> => {
    return this._createNode<Input, Output>(nodeProps, {
      parent,
      source: { name: 'external', nodeType: nodeProps.type },
    });
  };

  private _createNode = async <Input extends INodeOptions, Output extends INodeInstance>(
    nodeProps: Input,
    options: {
      parent?: INodeInstance;
      source: IGraphHook;
    }
  ): Promise<Output> => {
    const { parent, source } = options;

    // TODO: unique generation
    const id = nodeProps.id;

    const existing = this.getNode<Output>(id);
    if (existing) {
      console.warn(`Cannot add node. A node with id '${id}' already exists.`);
      return existing;
    }

    const type = nodeProps.type;
    if (source) {
      // intentionally forcing one node type per hook.
      // hooks should be small, with minimal concerns
      const allowedType = source.nodeType || source.name;

      if (source.nodeType && type !== allowedType) {
        console.error(
          `Creation of a node type not specified in the hook type is not allowed. Requested: ${type}; allowed: ${
            source.nodeType
          }`
        );
      }
    }

    // trigger the onWillCreateNode event
    this._hooks.forEach(hook => {
      // if hook not listening for event, don't run
      if (!hook.onWillCreateNode) return;

      // if not relevant to hook, don't run
      if (hook.name !== type && hook.nodeType !== type) return;

      try {
        // allow mutation here.. don't think this needs to be immutable
        hook.onWillCreateNode(nodeProps);
      } catch (e) {
        console.error('onWillCreateNode hook failed.', hook.name, e);
      }
    });

    const node = {
      // @ts-ignore some complicated TS bug here...
      ...nodeProps,
      id,
      type,

      // if their is a registered loader for this node type, call that, otherwise just return whatever content is
      loadContent: async () => {
        const loader = this._loaders[type];
        if (loader) {
          try {
            return loader(node);
          } catch (e) {
            console.error(`Error loading content for node ${node.id}`, e);
          }
        }

        return node.content;
      },
    } as Output;

    // add the node to our internal graph
    this.graph.setNode(id, node);

    // create the parent<->child relationship, if relevant
    if (parent) {
      this.graph.setParent(id, parent.id);
    }

    // trigger the onDidCreateNode event
    const didCreateEvents: Array<Promise<any>> = [];
    this._hooks.forEach(hook => {
      // if hook not listening for event, don't run
      if (!hook.onDidCreateNode) return;

      // no circular hooks if recursive false
      if (source && !source.recursive && source.name === hook.name) return;

      // if not relevant to hook, don't run
      let { selector } = hook;
      if (selector) {
        if (typeof selector !== 'function') {
          const nodeType = selector;
          selector = (n: INodeInstance) => n.type === nodeType;
        }
        if (!selector(node)) {
          return;
        }
      }

      // TODO(SL-248): the way hooks are designed is very limiting.
      // If a hook must wait until the WHOLE TREE is built it will be a problem for two reasons:
      // - we must ensure that plugins are ordered properly
      // - resolving node will currently fail with Promise.all
      didCreateEvents.push(
        hook.onDidCreateNode(node, {
          createNode: (n, opts) => this._createNode(n, { source: hook, ...opts }),
          createEdge: (sourceNode, destinatioNode) => this.graph.setEdge(sourceNode.id, destinatioNode.id),
          getNode: nodeId => this.getNode(nodeId),
          getParentNode: n => {
            const parentId = this.graph.parent(n.id);
            if (!parentId) {
              return;
            }
            return this.getNode(parentId);
          },
        })
      );
    });

    try {
      // TODO: is this the concurrency model we want? or more something like a queue, and don't block returning node?
      await Promise.all(didCreateEvents);
    } catch (e) {
      console.error('Some onDidCreateNode hooks failed.', e);
    }

    return node;
  };

  /**
   * Plugins
   */

  private _hooks: IGraphHook[] = [];
  private _loaders: {
    [nodeType: string]: IGraphLoader;
  } = {};

  public addPlugin = (...plugins: IGraphPlugin[]): Graph => {
    for (const plugin of plugins) {
      if (plugin.hooks) {
        for (const hook of plugin.hooks) {
          // Don't allow the same hook to be added multiple times
          const index = this._hooks.findIndex(h => h.name === hook.name);
          if (index >= 0) {
            console.warn(`Cannot add hook. A hook with name '${hook.name}' already exists.`);
            continue;
          }

          this._hooks.push(hook);
        }
      }

      if (plugin.loaders) {
        for (const nodeType in plugin.loaders) {
          if (!plugin.loaders.hasOwnProperty(nodeType)) continue;

          // Only one loader per nodeType allowed
          if (this._loaders[nodeType]) {
            console.warn(`Cannot add loader. A loader for node type '${nodeType}' already exists.`);
            continue;
          }

          this._loaders[nodeType] = plugin.loaders[nodeType];
        }
      }
    }
    return this;
  };
}
