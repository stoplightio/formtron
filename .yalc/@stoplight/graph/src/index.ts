export { Graph } from './graph';
export * from './types';
