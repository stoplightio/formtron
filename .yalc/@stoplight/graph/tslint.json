{
  "extends": ["./node_modules/@stoplight/scripts/tslint.json"]
}
