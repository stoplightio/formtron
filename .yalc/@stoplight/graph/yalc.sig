48b7f808e4ac24c4d54eed8754f3fa21