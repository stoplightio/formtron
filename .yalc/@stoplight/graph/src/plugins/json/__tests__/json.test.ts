import { Graph } from '../../../graph';
import { createJsonPlugin } from '../index';

describe('json plugin', () => {
  let graph: Graph;

  beforeEach(() => {
    graph = new Graph();
    graph.addPlugin(createJsonPlugin());
  });

  describe('onDidCreateNode', () => {
    test('should listen for new nodes with contentType json, and add a parsed child', async () => {
      await graph.createNode({
        id: '123',
        type: 'stringified-json',
        contentType: 'json',
        content: JSON.stringify({ foo: 'bar' }),
      });

      expect(graph.nodes[1].content).toMatchSnapshot();
    });

    test('should not add a parsed child for nodes that are not json', async () => {
      await graph.createNode({
        id: '123',
        type: 'text',
        contentType: 'text',
        content: 'foo',
      });

      expect(graph.nodes).toHaveLength(1);
    });
  });
});
