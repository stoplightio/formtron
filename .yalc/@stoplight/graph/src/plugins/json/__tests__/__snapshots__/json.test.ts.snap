// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`json plugin onDidCreateNode should listen for new nodes with contentType json, and add a parsed child 1`] = `
Object {
  "data": Object {
    "foo": "bar",
  },
  "pointers": Object {
    "": Object {
      "end": Object {
        "line": 1,
      },
      "start": Object {
        "line": 1,
      },
    },
    "/foo": Object {
      "end": Object {
        "line": 1,
      },
      "start": Object {
        "line": 1,
      },
    },
  },
  "validations": Array [],
}
`;
